[feature] spe_switch: Add spe_switch power feature

ActionHOSchT opened the pull request:
<!---
Describe what your pull request does,
i.e. fix this bug and how, add a feature, fix documentation…
If you add a feature, please answer these questions:
- what do you use the feature for?
- how does labgrid benefit as a testing library from the feature?
- how did you verify the feature works?
- if hardware is needed for the feature, which hardware is supported and which
  hardware did you test with?
--->
This PR adds the feature to set power for a phoenix_fl_switch over telnet.

Devices connected with single-pair-ethernet can now be set to power off/on/cycle.

To verify, this Driver was tested on a *FL SWITCH 2303-8SP1* with FW-version *3.27.01 BETA*.
Even if it's a BETA, we got the information that the power control won't experience breaking changes.

![grafik](https://github.com/JUMO-GmbH-Co-KG/labgrid/assets/73703991/16508302-0445-40c7-b691-9a023aabbe0b)
![grafik](https://github.com/JUMO-GmbH-Co-KG/labgrid/assets/73703991/1a910d49-a660-4d41-8fe6-3c1513d5473b)

![grafik](https://github.com/JUMO-GmbH-Co-KG/labgrid/assets/73703991/1eee564d-3c13-44f7-b058-42abf95d48b4)

<!---
This checklist roughly outlines the steps for new features, remove and add tasks as needed:
--->
**Checklist**
- [x] Documentation for the feature
- [x] Tests for the feature 
<!---
If you add a driver/resource or modify one:
--->
- [x] The arguments and description in doc/configuration.rst have been updated
<!---
Did you test the change locally? If yes, best to mention how you did it in the description section.
--->
- [x] PR has been tested

ActionHOSchT commented:
@EugenWiens @ChHapp @th-fischer2020 @a-kipp 
Please review.

EugenWiens commented on doc/configuration.rst:
```diff
@@ -233,6 +233,41 @@ Currently available are:
+.. code-block:: yaml
+
+   NetworkPowerPort:
+     model: spe_switch
```
I think the model `spe_switch` is not precise enough. The Telnet API ist he API of the FL SWITCH 2303-8SP other manufacturer of spe switch device can have an other API.
Please rename the model

ActionHOSchT replied:
Renamed to *phoenix_fl_switch*

ChHapp commented on labgrid/driver/powerdriver.py:
```diff
@@ -1,16 +1,16 @@
+import math
 import shlex
 import time
-import math
```
Why do you have to change the import order of the existing imports?

ActionHOSchT replied:
The contribution-guideline says, that one should always use isort to sort the imports. It seems that the imports weren't sorted.

ActionHOSchT replied:
outdated, changes reverted

ChHapp commented on labgrid/driver/powerdriver.py:
```diff
@@ -232,6 +232,111 @@ def cycle(self):
+
+@target_factory.reg_driver
+@attr.s(eq=False)
+class NetworkPowerDriverWithAuth(Driver, PowerResetMixin, PowerProtocol):
```
Why do you need a further `NetworkPowerDriverWithAuth` function here? How does the netio driver works?
As far as I can remember there is also creds necessary by using the web-api.

ActionHOSchT replied:
The netio json-api doesn't need credentials if it's not forced.
We as an user could only use default-credentials or stick the custom credentials to an url.
Why would we parse the Host to extract our user and password, if we could simply provide this as own sub-asset for our resource and process them in our drivers. For example *phoenix_fl_switch* needs a user and password to login via telnet.

ChHapp replied:
That might be so for the json-api but if you decide to use the web-api (https://github.com/JUMO-GmbH-Co-KG/labgrid/blob/master/labgrid/driver/power/netio.py#L15) the creds are hardcoded in the driver. 
However, `class NetworkPowerDriverWithAuth(Driver, PowerResetMixin, PowerProtocol):` - your new `class` has the same parameters as the existing `class NetworkPowerDriver(Driver, PowerResetMixin, PowerProtocol):` 
My suggestion would be to use in the first place a given (hardcoded creds like the netio-driver (admin:admin) does), further in a separate MR (step) it might be better to extends just the existing `class NetworkPowerDriver(Driver, PowerResetMixin, PowerProtocol):` with the missing members for `username` and `password`. At the first glance it seems you have a lot of duplicated code by creating your separate `class`.

ActionHOSchT replied:
Everything which was introduced with `class NetworkPowerDriverWithAuth` is reverted.
The std-creds were added and it will no longer be possible to set custom credentials. However a future rework with custom creds should be pretty easy.

ChHapp commented on .vscode/settings.json:
```diff
@@ -0,0 +1,44 @@
+{
```
Remove this file please, I guess it was added by accident, thx

ActionHOSchT replied:
Yes, this was added by accident.
to prevent project-settings from being accidently added `.vscode` was added in the git ignore (right below `.idea`)

ActionHOSchT committed:
[feature] add phoenix_fl_switch

added single-pair-ethernet power support over telnet

changed name as requested in https://github.com/JUMO-GmbH-Co-KG/labgrid/pull/4#discussion_r1310190801

set default for username and password

added .vscode to .gitignore

Signed-off-by: Raffael Krakau <r.krakau90@gmail.com>

ActionHOSchT committed:
[feature] add doc

Signed-off-by: Raffael Krakau <r.krakau90@gmail.com>

EugenWiens commented on labgrid/driver/power/phoenix_fl_switch.py:
```diff
@@ -0,0 +1,80 @@
+        - disable(False): Turn OFF,
+        - enable(True): Turn ON
+    """
+    username = "admin"
+    password = "private"
```
is it possible to move the credentials to a getter ?  The username and password is defined in both functions. May be it is a better idea to create a login function.

ActionHOSchT replied:
imo any work on the credentials is superfluous, since it will be replaced asap

Hygdrasil commented on labgrid/driver/power/phoenix_fl_switch.py:
```diff
@@ -0,0 +1,80 @@
+    if "OK" in status:
+        pass
+    else:
+        print("Could not set power")
+        raise Exception
```
Why is there a separate printout before the exception?
```suggestion
        raise Exception("Could not set power")
```

ActionHOSchT replied:
this should be as you mentioned. Will be adjusted.

Hygdrasil commented on labgrid/driver/power/phoenix_fl_switch.py:
```diff
@@ -0,0 +1,80 @@
+    telnet = Telnet(host=host, port=port)
+
+    # login user with password
+    telnet.read_until(match=b'User: ')
+    telnet.write(bytes(f'{username}\r\n', "utf-8"))
+    telnet.read_until(match=b'Password: ')
+    telnet.write(bytes(f'{password}\r\n', "utf-8"))
```
This is duplicated. Please move this section to a separate function.

ActionHOSchT replied:
This is oriented at `netio_kshell`, which also uses telnet. Therefore it seems to have to be duplicated.
https://github.com/JUMO-GmbH-Co-KG/labgrid/blob/master/labgrid/driver/power/netio_kshell.py

ChHapp commented on labgrid/driver/power/phoenix_fl_switch.py:
```diff
@@ -0,0 +1,80 @@
+Copyright 2023 JUMO GmbH & Co. KG
+'''
+
+from telnetlib import Telnet
```
The build node currently has Python 3.10.12 - 
```
 Deprecated since version 3.11, will be removed in version 3.13: The [telnetlib](https://docs.python.org/3/library/telnetlib.html#module-telnetlib) module is deprecated (see [PEP 594](https://peps.python.org/pep-0594/#telnetlib) for details and alternatives).
 ```

Have a look at https://pypi.org/project/telnetlib3/ maybe an alternative for that  :thinking:

ActionHOSchT replied:
switched to pexpect.spawn(…"telnet…")

ChHapp commented on labgrid/driver/power/phoenix_fl_switch.py:
```diff
@@ -0,0 +1,80 @@
+    action = "enable" if value else "disable"
+
+    telnet = Telnet(host=host, port=port)
+
```
I would suggest to use the follow lines with the `with` statement: 
 (https://docs.python.org/3/library/telnetlib.html)
```
with Telnet('localhost', 23) as tn:
    tn....
    tn...
    . 
    . 


```
further it might be good to have it encapsulated in a helper function

ActionHOSchT replied:
besides using pexpect done

ChHapp replied:
:+1:

ChHapp commented on labgrid/driver/power/phoenix_fl_switch.py:
```diff
@@ -0,0 +1,80 @@
+    telnet.close()
+    # check status
+    if "OK" in status:
+        status = status.split("Status :  ")[1].splitlines()[0]
```
just check if the access on index `[1]` or `[0]` could not go wrong - a better solution would be without those

ActionHOSchT replied:
Reworked due to new/convenient possibilities with pexpect.

ActionHOSchT committed:
[review] use with pexpect.spawn

reworked telnet-handling
moved login to function

Signed-off-by: Raffael Krakau <r.krakau90@gmail.com>

ActionHOSchT commented:
@Hygdrasil @EugenWiens @ChHapp  Login is now separated into own function. pls review

EugenWiens commented on labgrid/driver/power/phoenix_fl_switch.py:
```diff
@@ -0,0 +1,76 @@
+    tn.send(bytes(f'{username}\r\n', "utf-8"))
+    tn.expect(b'Password: ')
+    tn.send(bytes(f'{password}\r\n', "utf-8"))
+    return tn
```
returning the same object is not necessary, phyton use call by ref. That means you have changed the origin object

ActionHOSchT replied:
thank you for advice/explanation

changed and force pushed to keep history clean. (this is the reason this comment isn't mentioned in commit)

EugenWiens commented on labgrid/driver/power/phoenix_fl_switch.py:
```diff
@@ -0,0 +1,75 @@
+    Login user with set credentials
+
+    @param tn : pyexpect-telnet-object
+    @returns : logged in pyexpect-telnet-object
```
```suggestion
```

ChHapp commented on labgrid/driver/power/phoenix_fl_switch.py:
```diff
@@ -0,0 +1,75 @@
+PORT = 23
+
+
+def login_telnet(tn):
```
```suggestion
def __login_telnet(tn):
```

ActionHOSchT committed:
[review] apply suggestions from https://github.com/JUMO-GmbH-Co-KG/labgrid/pull/4
